[libc] Lazily initialize freelist malloc using symbols

mysterymath opened the pull request:
This requires the user to set the upper bounds of the heap by defining the symbol `__libc_heap_limit`. The heap begins at `_end` and ends `__libc_heap_limit` bytes afterwards. This prevents a completely unused heap from requiring any space, and it prevents the heap from being zeroed at initialization time as part of BSS. It also allows users to customize the available heap location without recompiling libc.
    
I'd think this should eventually be replaced with an implemenation based on a morecore() library. This would allow the same implementation to use sbrk() on POSIX, `_end` and `__libc_heap_limit` on embedded, and a buffer in tests. It would also provide better "wilderness" behavior that tends to decrease heap fragementation (see Wilson et al.)

See #98096

mysterymath committed:
[libc] Lazily initialize freelist malloc using symbols

This allows the user to customize the size of the heap provided by
overriding the weak symbol __libc_heap_limit at link time. The heap
begins at _end and ends __libc_heap_limit bytes afterwards. It also
prevents a completely unused heap from requiring any space. (It's
reasonably common to link in malloc/free to handle exceptional
situations that can never actually be encountered.)

I'd think this should eventually be replaced with an implemenation based
on sbrk(), with sbrk() implemented generically in terms of brk(), which
would be a system call on e.g. Linux and a bump pointer from _end up to
__libc_heap_limit on baremetal. This would allow the allocator to be
more flexible (e.g., an implementation could swap out brk() and do
dynamic memory availability checks), to manage the heap better by
keeping better track of "wilderness" and to work properly as a malloc on
Linux.

Incidentally, this sets the default heap limit to 128KiB, from 102400
bytes, which had been reported as "1GiB".

petrhosek commented on libc/src/stdlib/heap_limit.S:
```diff
@@ -0,0 +1,2 @@
+.weak __libc_heap_limit
+.set __libc_heap_limit, LIBC_FREELIST_MALLOC_SIZE
```
Can you define this in `freelist_malloc.cpp`? We want to avoid any unnecessary assembly in libc.

mysterymath replied:
I dodged this entirely by moving to `__libc_heap_limit` to be an address rather than a size; this removes `LIBC_FREELIST_MALLOC_SIZE` entirely. `__libc_heap_limit` would then be an intentionally undefined reference in the libc.

petrhosek commented on libc/src/__support/freelist_heap.h:
```diff
@@ -20,6 +20,9 @@
 #include "src/string/memory_utils/inline_memcpy.h"
 #include "src/string/memory_utils/inline_memset.h"
 
+extern char _end;
+extern char __libc_heap_limit;
```
```suggestion
extern "C" char _end;
extern "C" char __libc_heap_limit;
```

mysterymath committed:
Extern "C" in namespace for heap symbols

PiJoules commented on libc/src/__support/freelist_heap.h:
```diff
@@ -47,17 +50,10 @@ template <size_t NUM_BUCKETS = DEFAULT_BUCKETS.size()> class FreeListHeap {
-        freelist_(DEFAULT_BUCKETS), heap_stats_{} {
-    heap_stats_.total_bytes = total_bytes;
+  constexpr FreeListHeap(span<cpp::byte> region)
+      : is_initialized_(false), region_(region), freelist_(DEFAULT_BUCKETS),
```
Not sure if I'm overthinking this, but are these guaranteed to be link-initialized to their default values for the global in freelist_malloc.cpp?

PiJoules commented on libc/src/stdlib/freelist_malloc.cpp:
```diff
@@ -19,16 +19,13 @@
-// This is set via the LIBC_CONF_FREELIST_MALLOC_BUFFER_SIZE configuration.
-constexpr size_t SIZE = LIBC_FREELIST_MALLOC_SIZE;
-#else
+#ifndef LIBC_FREELIST_MALLOC_SIZE
```
We can probably remove this here since it's unused here.

mysterymath committed:
Address comments:

- Move preprocessor check to assembly file
- Add LIBC_CONSTINIT back to heap definition; make non-constexpr cast
  from &__heap_limit to size_t occur at init time to restore constexpr
  init.

mysterymath committed:
Default initializers to clean up constructors

mysterymath committed:
Make __libc_heap_limit an end symbol, not a size

This is more conventional and it dodges the issue of needing to provide
a default size entirely. Setting up a baremetal heap intrinsically
requires some linker script work, and __libc_heap_limit will be an
undefined reference hinting to the user that this work must be done.

Once we have a morecore function, a baremetal morecore() would use
__libc_heap_limit, a Linux one would use sbrk(), and a test one would
use a buffer.

petrhosek commented on libc/src/__support/freelist_heap.h:
```diff
@@ -22,6 +22,9 @@
 
 namespace LIBC_NAMESPACE_DECL {
 
+extern "C" cpp::byte _end;
```
I'd prefer using C types for `extern "C"` symbols so users don't have to guess what C type does this correspond to.

mysterymath replied:
I'd really like `extern "C" void _end;` to be legal for cases like this, but it isn't. There's no correct C type, since there is no C object that this symbol refers to. It's instead a GCC-blessed abuse of notation to obtain the symbol value via `&_end`. (I'm not sure whether or not this technically violates the standard; hadn't occurred to me before this reply.)

I made it a `cpp::byte` only because `cpp::byte*` seems to have become the preferred type for pointers to undifferentiated memory. The more conventional way to do this is `extern "C" char _end;`, but that might only be because this pattern appears far more commonly in C than in C++.

petrhosek commented on libc/src/__support/freelist_heap.h:
```diff
@@ -22,6 +22,9 @@
 namespace LIBC_NAMESPACE_DECL {
 
+extern "C" cpp::byte _end;
+extern "C" cpp::byte __libc_heap_limit;
```
Is `__libc_heap_limit` already used by other C libraries (in other words is this an existing API)? If not, I'd prefer calling it `__llvm_libc_heap_limit` which matches the naming scheme we use for other symbols that are part of the external ABI.

mysterymath replied:
Ah, that was my mistake about the convention. Done.

mysterymath committed:
Address review comments

- __libc_heap_limit -> __llvm_libc_heap_limit

PiJoules commented on libc/test/src/__support/heap_limit_fake.cpp:
```diff
@@ -0,0 +1,20 @@
+extern "C" {
+// This isn't used in the unit tests, but it must be defined for the non-fake
+// version of the heap to work.
+cpp::byte __llvm_libc_heap_limit;
```
Does `_end` also need to be defined here?

mysterymath replied:
Hmm yeah. I had been relying on the linker-generated version on Linux, which is how I'd been locally running these texts (patching in the entrypoint). Done.

mysterymath committed:
Include fake for _end too.

PiJoules commented on libc/test/src/__support/freelist_heap_test.cpp:
```diff
@@ -30,14 +30,19 @@ using LIBC_NAMESPACE::freelist_heap;
 #define TEST_FOR_EACH_ALLOCATOR(TestCase, BufferSize)                          \
   class LlvmLibcFreeListHeapTest##TestCase : public testing::Test {            \
   public:                                                                      \
+    FreeListHeapBuffer<BufferSize> fake_global_buffer;                         \
+    void SetUp() override {                                                    \
+      freelist_heap =                                                          \
+          new (&fake_global_buffer) FreeListHeapBuffer<BufferSize>;            \
+    }                                                                          \
```
Since we have `heap_limit_fake.cpp`, is it be possible to define `_end` and `__llvm_libc_heap_limit` to be the start and end of some static global buffer so we can assert the linker-initialized `freelist_heap` is WAI?

mysterymath replied:
Not for these tests without limitation, since the buffer size is variable, but that does work for the malloc test, which gives us some smoke testing that the symbol version works. That's probably most of what's needed for now; the `morecore` variant should make this more straightfoward to test later. Done.

mysterymath committed:
Define _end and __llvm_libc_heap_limit to fake buffer for malloc test

PiJoules reviewed (approved):
LGTM. Should probably get a libc owner to also approve.

mysterymath committed:
Clang format

mysterymath commented:
@michaelrj-google @lntue, is there anything left to do on this one?

lntue commented:
> @michaelrj-google @lntue, is there anything left to do on this one?

no, I was just waiting for you to sync the PR to HEAD and the presubmit bot to finish.
